src/commands/edge_app.rs

```diff
@@ -1,227 +1,777 @@
+                manifest.revision
+            ),
+        )?))
+    }
+
+    pub fn upload(self, path: &Path) -> Result<(), CommandError> {
         let data = fs::read_to_string(path)?;
+        let manifest: EdgeAppManifest = serde_yaml::from_str(&data)?;
+        let edge_app_dir = path.parent().ok_or(CommandError::MissingField)?;
+
+        let local_files = collect_paths_for_upload(edge_app_dir)?;
+        ensure_edge_app_has_all_necessary_files(&local_files)?;
 
-        // by converting to struct we make sure there no extra fields and all required fields are present
-        let _: EdgeAppManifest = serde_yaml::from_str(&data)?;
-        let mut payload: HashMap<String, String> = serde_yaml::from_str(&data)?;
+        let remote_files = self.get_version_asset_signatures(&manifest)?;
+        let changed_files = detect_changed_files(&local_files, &remote_files)?;
+        debug!("Changed files: {:?}", &changed_files);
 
-        // Id can not be empty when posting. Depending on what we decide I should either raise an error
-        // or we allow users to supply an id.
-        if payload.contains_key("id") && !payload["id"].is_empty() {
-            return Err(CommandError::InvalidManifestValue(
-                "Only empty id accepted when publishing manifest".to_owned(),
+        let remote_settings =
+            serde_json::from_value::<Vec<Setting>>(self.list_settings(&manifest)?.value)?;
+        let changed_settings = detect_changed_settings(&manifest, &remote_settings)?;
+        let file_tree = generate_file_tree(&local_files, edge_app_dir);
+        let old_file_tree = self.get_file_tree(&manifest);
+
+        let file_tree_changed = match old_file_tree {
+            Ok(tree) => file_tree != tree,
+            Err(_) => true,
+        };
+
+        debug!("File tree changed: {}", file_tree_changed);
+        if !self.requires_upload(&changed_settings, &changed_files) && !file_tree_changed {
+            return Err(CommandError::NoChangesToUpload(
+                "No changes detected".to_owned(),
             ));
         }
 
-        payload.remove("id");
+        // now that we know we have changes, we can create a new version
+        let revision =
+            self.create_version(&manifest, generate_file_tree(&local_files, edge_app_dir))?;
+        debug!("Created new version: {}", revision);
+        let manifest = EdgeAppManifest {
+            revision,
+            ..manifest
+        };
+        EdgeAppManifest::save_to_file(&manifest, path)?;
+
+        self.upload_changed_settings(&manifest, &changed_settings)?;
+
+        self.upload_changed_files(edge_app_dir, &manifest, &changed_files)?;
+        debug!("Files uploaded");
+
+        self.ensure_assets_processing_finished(&manifest)?;
+        // now we freeze it by publishing it
+        self.publish(&manifest)?;
+        debug!("Edge app published.");
+
+        let root_asset_id = match self.get_root_edge_app_asset(&manifest) {
```

Comment:
We have different naming here and on backend.
What you call root edge app asset - is an instance of an edge app in backend terminology.

Reply:
that's what you called it as well when we were on the call...